File: src/dots/core/schema.py
"""Schema validation for path.yaml dependencies and files."""

from typing import Optional


# Campos v2 que indican schema obsoleto
V2_DEP_FIELDS = {"source", "target", "extract-path", "arch_map", "package-managers"}
V2_FILE_FIELDS = {"destination-linux", "destination-mac", "destination-override"}


def detect_v2_schema(data: dict, yaml_path: str) -> list[str]:
    """
    Detecta uso de schema v2 y retorna error early fail.
    Se ejecuta antes de parsear para dar mensaje claro.
    """
    errors = []

    # Check dependencies
    deps = data.get("dependencies", [])
    for i, dep in enumerate(deps):
        if isinstance(dep, dict) and V2_DEP_FIELDS.intersection(dep.keys()):
            errors.append(
                f"Schema v2 detected in dependencies[{i}] ({yaml_path}). "
                f"Run 'dots migrate' to upgrade to v3 automatically."
            )

    # Check files
    files = data.get("files", [])
    for i, f in enumerate(files):
        if isinstance(f, dict) and V2_FILE_FIELDS.intersection(f.keys()):
            errors.append(
                f"Schema v2 detected in files[{i}] ({yaml_path}). "
                f"Run 'dots migrate' to upgrade to v3 automatically."
            )

    return errors


# Campos requeridos por tipo de dependency
REQUIRED_FIELDS: dict[str, list[str]] = {
    "binary": ["url", "dest"],
    "git": ["url", "dest"],
    "package": ["name"],
}


def validate_dependency(raw: dict, yaml_path: str) -> list[str]:
    """
    Valida un dict crudo de dependency.
    Retorna lista de errores (vacía si todo ok).
    """
    errors = []
    dep_type = raw.get("type", "package")
    dep_name = raw.get("name", "<unnamed>")
    prefix = f"[{yaml_path}] dependency '{dep_name}'"

    # Tipo válido
    if dep_type not in REQUIRED_FIELDS:
        errors.append(f"{prefix}: type '{dep_type}' desconocido (conocidos: {', '.join(REQUIRED_FIELDS.keys())})")

    # Campos requeridos por tipo
    if dep_type in REQUIRED_FIELDS:
        for field in REQUIRED_FIELDS[dep_type]:
            if not raw.get(field):
                errors.append(f"{prefix}: campo requerido '{field}' faltante para type '{dep_type}'")

    return errors


def validate_file_mapping(raw: dict, yaml_path: str) -> list[str]:
    """
    Valida un dict crudo de file mapping.
    Retorna lista de errores (vacía si todo ok).
    """
    errors = []
    source = raw.get("source", "<unnamed>")
    prefix = f"[{yaml_path}] file mapping '{source}'"

    if not raw.get("source"):
        errors.append(f"{prefix}: sin 'source'")

    has_destination = raw.get("destination") or raw.get("per-os")
    if not has_destination:
        errors.append(f"{prefix}: sin 'destination' ni 'per-os'")

    # Validar per-os si existe
    per_os = raw.get("per-os")
    if per_os and not isinstance(per_os, dict):
        errors.append(f"{prefix}: 'per-os' debe ser un dict")

    # Validar os si existe
    os_filter = raw.get("os")
    if os_filter and not isinstance(os_filter, list):
        errors.append(f"{prefix}: 'os' debe ser una lista")

    return errors


def validate_path_yaml(data: dict, yaml_path: str) -> list[str]:
    """
    Valida un path.yaml completo.
    Retorna lista de errores (vacía si todo ok).
    """
    errors = []

    if not isinstance(data, dict):
        return [f"[{yaml_path}]: debe ser un dict"]

    # Validar dependencies
    dependencies = data.get("dependencies", [])
    if dependencies and not isinstance(dependencies, list):
        errors.append(f"[{yaml_path}]: 'dependencies' debe ser una lista")
    elif isinstance(dependencies, list):
        for i, dep in enumerate(dependencies):
            if isinstance(dep, dict):
                errors.extend(validate_dependency(dep, yaml_path))
            elif isinstance(dep, str):
                # Strings son válidas (legacy shorthand)
                pass
            else:
                errors.append(f"[{yaml_path}]: dependency #{i} debe ser dict o string")

    # Validar files
    files = data.get("files", [])
    if files and not isinstance(files, list):
        errors.append(f"[{yaml_path}]: 'files' debe ser una lista")
    elif isinstance(files, list):
        for i, file_map in enumerate(files):
            if isinstance(file_map, dict):
                errors.extend(validate_file_mapping(file_map, yaml_path))
            else:
                errors.append(f"[{yaml_path}]: file #{i} debe ser un dict")

    return errors

File: internal/yaml/schema.go
package yaml

import "fmt"

// V2DepFields are fields that indicate an obsolete v2 schema for dependencies.
var V2DepFields = map[string]bool{
	"source": true, "target": true, "extract-path": true,
	"arch_map": true, "package-managers": true,
}

// V2FileFields are fields that indicate an obsolete v2 schema for files.
var V2FileFields = map[string]bool{
	"destination-linux": true, "destination-mac": true,
	"destination-override": true,
}

// RequiredFieldsByType maps dependency types to their required fields.
var RequiredFieldsByType = map[string][]string{
	"binary":  {"url", "dest"},
	"git":     {"url", "dest"},
	"package": {"name"},
}

// DetectV2Schema checks if a YAML data dict uses v2 schema.
func DetectV2Schema(data map[string]interface{}, yamlPath string) []string {
	var errors []string

	// Check dependencies
	if deps, ok := data["dependencies"].([]interface{}); ok {
		for i, dep := range deps {
			if d, ok := dep.(map[string]interface{}); ok {
				for field := range d {
					if V2DepFields[field] {
						errors = append(errors, fmt.Sprintf(
							"Schema v2 detected in dependencies[%d] (%s). Run 'dots migrate' to upgrade to v3 automatically.",
							i, yamlPath,
						))
						break
					}
				}
			}
		}
	}

	// Check files
	if files, ok := data["files"].([]interface{}); ok {
		for i, f := range files {
			if file, ok := f.(map[string]interface{}); ok {
				for field := range file {
					if V2FileFields[field] {
						errors = append(errors, fmt.Sprintf(
							"Schema v2 detected in files[%d] (%s). Run 'dots migrate' to upgrade to v3 automatically.",
							i, yamlPath,
						))
						break
					}
				}
			}
		}
	}

	return errors
}

// ValidateDependency validates a single dependency dict.
func ValidateDependency(raw map[string]interface{}, yamlPath string) []string {
	var errors []string
	depType, _ := raw["type"].(string)
	if depType == "" {
		depType = "package"
	}
	depName, _ := raw["name"].(string)
	if depName == "" {
		depName = "<unnamed>"
	}
	prefix := fmt.Sprintf("[%s] dependency '%s'", yamlPath, depName)

	// Valid type
	if _, ok := RequiredFieldsByType[depType]; !ok {
		knownTypes := make([]string, 0, len(RequiredFieldsByType))
		for t := range RequiredFieldsByType {
			knownTypes = append(knownTypes, t)
		}
		errors = append(errors, fmt.Sprintf(
			"%s: type '%s' unknown (known: %v)", prefix, depType, knownTypes,
		))
	}

	// Required fields by type
	if fields, ok := RequiredFieldsByType[depType]; ok {
		for _, field := range fields {
			if v, _ := raw[field].(string); v == "" {
				errors = append(errors, fmt.Sprintf(
					"%s: required field '%s' missing for type '%s'", prefix, field, depType,
				))
			}
		}
	}

	return errors
}

// ValidateFileMapping validates a single file mapping dict.
func ValidateFileMapping(raw map[string]interface{}, yamlPath string) []string {
	var errors []string
	source, _ := raw["source"].(string)
	if source == "" {
		source = "<unnamed>"
	}
	prefix := fmt.Sprintf("[%s] file mapping '%s'", yamlPath, source)

	if raw["source"] == nil || raw["source"].(string) == "" {
		errors = append(errors, fmt.Sprintf("%s: missing 'source'", prefix))
	}

	_, hasDest := raw["destination"].(string)
	_, hasPerOS := raw["per-os"].(map[string]interface{})
	if !hasDest && !hasPerOS {
		errors = append(errors, fmt.Sprintf("%s: no 'destination' or 'per-os'", prefix))
	}

	if perOS, ok := raw["per-os"]; ok {
		if _, ok := perOS.(map[string]interface{}); !ok {
			errors = append(errors, fmt.Sprintf("%s: 'per-os' must be a dict", prefix))
		}
	}

	if osFilter, ok := raw["os"]; ok {
		if _, ok := osFilter.([]interface{}); !ok {
			errors = append(errors, fmt.Sprintf("%s: 'os' must be a list", prefix))
		}
	}

	return errors
}

// ValidatePathYAML validates a complete path.yaml structure.
func ValidatePathYAML(data map[string]interface{}, yamlPath string) []string {
	var errors []string

	if data == nil {
		return []string{fmt.Sprintf("[%s]: must be a dict", yamlPath)}
	}

	// Validate dependencies
	if deps, ok := data["dependencies"].([]interface{}); ok {
		for _, dep := range deps {
			if d, ok := dep.(map[string]interface{}); ok {
				errors = append(errors, ValidateDependency(d, yamlPath)...)
			}
		}
	}

	// Validate files
	if files, ok := data["files"].([]interface{}); ok {
		for _, f := range files {
			if file, ok := f.(map[string]interface{}); ok {
				errors = append(errors, ValidateFileMapping(file, yamlPath)...)
			}
		}
	}

	return errors
}
